src/main/java/seedu/address/model/task/TaskList.java

```diff
@@ -82,6 +96,56 @@ public void remove(Task toRemove) {
         }
     }
 
+    /**
+     * If a recurring task's original DateTime is before the current {@code DateTime},
+     * changes the task's {@code DateTime} to the next earliest {@code DateTime},
+     * according to whether is it a {@code DAY}, {@code WEEK} or {@code MONTH} recurrence type.
+     * At the same time, after changing the {@code DateTime},
+     * it also maintains the sorted order of tasks according to {@code DateTime}.
+     * All tasks which had their {@code DateTime} changed will also have their status's {@code isDone} as false.
+     */
+    public void changeDateOfPastRecurringTasks() {
+        LocalDateTime currentDateTime = LocalDateTime.now();
+
+        for (Task t : this.internalList) {
+            Recurrence recurrence = t.getRecurrence();
+            if (recurrence.isRecurring()) {
+                RecurrenceType recurrenceType = recurrence.getRecurrenceType();
+                assert(recurrenceType != RecurrenceType.NONE);
+                DateTime taskDateTime = t.getDateTime();
+                LocalDate taskDate = taskDateTime.date;
+                LocalTime taskTime = taskDateTime.time;
+                DateTimeFormatter dtf = DateTimeFormatter.ofPattern("dd MM yyyy HH mm");
+                String inputString = String.format("%02d", taskDate.getDayOfMonth()) + " "
+                        + String.format("%02d", taskDate.getMonthValue()) + " "
+                        + String.format("%02d", taskDate.getYear()) + " "
+                        + String.format("%02d", taskTime.getHour()) + " "
+                        + String.format("%02d", taskTime.getMinute());
+                LocalDateTime taskLocalDateTime = LocalDateTime.parse(inputString, dtf);
+                long daysBetween = Duration.between(taskLocalDateTime, currentDateTime).toDays();
+                // if task dateTime is before current dateTime
+                if (daysBetween > 0) {
+                    LocalDateTime taskNewLocalDateTime;
+                    if (recurrenceType == RecurrenceType.DAY) {
+                        taskNewLocalDateTime = taskLocalDateTime.plusDays(daysBetween + 1);
+                    } else if (recurrenceType == RecurrenceType.WEEK) {
+                        int daysToAdd = (int) (daysBetween / 7) * 7 + 7;
+                        taskNewLocalDateTime = taskLocalDateTime.plusDays(daysToAdd);
+                    } else {
+                        // assume its + 4 weeks
+                        int daysToAdd = (int) (daysBetween / 28) * 28 + 28;
+                        taskNewLocalDateTime = taskLocalDateTime.plusDays(daysToAdd);
+                    }
+                    // time is fixed
+                    t.setDateTime(
+                            new DateTime(taskNewLocalDateTime.toLocalDate(), taskLocalDateTime.toLocalTime()));
+                }
```

Comment:
Perhaps this section could be abstracted out as it's own method for better code quality? It seems like the method has a deep arrowhead structure right now